### PR TITLE
Bnb/tf min

### PR DESCRIPTION
**Pull request by bnb32**
Minimal changes for compatibility with the tensorflow 2.16+ compliant upgrade of phygnn.
* Mostly in `abstract.py` and centered around preserving parallel_grad functionality while dealing with API changes.
* Lots of `__call__` to `call` changes for loss metrics.
* Additional minor changes to weight initialization and optimizer handling
* pixi.lock changes are most of the line count

### CODE REVIEW

File: sup3r/models/abstract.py [hunk 1]
```diff
@@ -320,6 +319,47 @@ def generator_weights(self):
         """
         return self.generator.weights
 
+    def _init_generator_weights(self, lr_shape, hr_shape, device=None):
```
Comment (bnb32) [inline sup3r/models/abstract.py:332]: ^ weight init used for conditional models also

File: sup3r/models/abstract.py [hunk 2]
```diff
@@ -339,13 +379,10 @@ def init_optimizer(optimizer, learning_rate):
...
         """
         if isinstance(optimizer, dict):
-            class_name = optimizer['name']
-            optimizer_class = getattr(optimizers, class_name)
-            sig = signature(optimizer_class)
-            optimizer_kwargs = {
-                k: v for k, v in optimizer.items() if k in sig.parameters
-            }
-            optimizer = optimizer_class.from_config(optimizer_kwargs)
+            optimizer = optimizers.deserialize({
```
Comment (bnb32) [inline sup3r/models/abstract.py:392]: this is more reliable in keras 3

File: sup3r/models/abstract.py [hunk 3]
```diff
@@ -1266,34 +1328,54 @@ def _tf_generate(self, low_res, hi_res_exo=None):
...
+            'be False. Received train_gen={} and train_disc={}'.format(
+                train_gen, train_disc
+            )
+        )
+        logger.error(msg)
+        raise ValueError(msg)
+
     @tf.function
-    def _tf_get_single_grad(
+    def get_single_grad_gen(
```
Comment (bnb32) [inline sup3r/models/abstract.py:1347]: A fair bit changed in this file. Passing training weights through as args is fragile in keras 3. Went back and forth a bunch on best approach for maintaining tf.function coverage and training speed.

File: sup3r/models/base.py [hunk 4]
```diff
@@ -354,14 +382,14 @@ def update_optimizer(self, option='generator', **kwargs):
         if 'gen' in option.lower() or 'all' in option.lower():
             conf = self.get_optimizer_config(self.optimizer)
             conf.update(**kwargs)
-            optimizer_class = get_optimizer_class(conf)
-            self._optimizer = optimizer_class.from_config(conf)
+            self._optimizer = self.optimizer.__class__.from_config(conf)
```
Comment (bnb32) [inline sup3r/models/base.py]: keras 3 driven change

File: sup3r/utilities/loss_metrics.py [hunk 5]
```diff
@@ -127,7 +127,8 @@ def gaussian_kernel(x_true, x_gen, sigma=1.0):
 class ExpLoss(Sup3rLoss):
     """Loss class for squared exponential difference"""
 
-    def __call__(self, x_true, x_gen):
+    @staticmethod
+    def call(x_true, x_gen):
```
Comment (bnb32) [inline sup3r/utilities/loss_metrics.py:131]: changed `__call__` to `call` everywhere. This seems to be the standard pattern.

**Comment by bnb32** [comment]
@grantbuster mostly same sentiment as https://github.com/NatLabRockies/phygnn/pull/72#issuecomment-4309719047
